CMS default fields removal configuration

mfendeksilverstripe opened the pull request:
# CMS default fields

Scaffolded fields are nice for a quick project startups but are often hindrance as these fields make customisation of CMS fields more complex. Most notable cases are:

* customising existing scaffolded field takes more lines of code compared to defining it from scratch
* customising field order takes more lines of code compared to defining it from scratch

In many cases it's preferable to simply remove these scaffolded fields and redefine them in the model. As a result field definition is more readable as all fields are defined in one place. The code is also easier to maintain as we don't have to be concerned on where in the `FieldList` the scaffolded fields are located.

## Configure which scaffolded fields are kept and which are removed

I couldn't avoid creating scaffolded fields as there are many code dependencies which assume scaffolded fields will be there. This new extension allows configurable removal of scaffolded fields. The moment of removal is also configurable as this extension is not applied by default, it has to be applied manually.

## Compatibility

This feature is disabled by default to enable it you have to:

* apply extension to a data object
* provide field removal configuration (empty configuration does nothing)

## Related issues

https://github.com/silverstripe/silverstripe-framework/issues/9403

michalkleiner commented:
As much as I like the idea, because we're often in the same situation of removing or replacing the default scaffolded fields, I think this approach is super convoluted and perhaps overly complicated. Don't have much more time to provide an alternative solution, though simply "turn off default form fields scaffolding" would be sufficient for starters.

A couple of points to the code itself:
- even though the classname is namespaced, it's a good practice to still include some basic keyword in the classname about what it does, not just a generic "extension", as when used through the class name itself without the FQDN or without an alias, it's confusing. I'd suggest something along the lines of DisableScaffoldingExtension or similar. It's longer, but more descriptive.
I see you prefer this pattern through multiple PR's, perhaps a documented guidance in the Contributing docs from the core team would be helpful?
- mix of snake_case, kebab-case and camelCase config options.

mfendeksilverstripe commented:
@michalkleiner Thank you for your feedback.

Just to clarify

```
field_removal:
  db-keep: # remove all db fields except Title and ShowTitle
    fields:
      Title: true
```

`field_removal` is a configuration API property which is supposed to be snake_case
`db-keep` is an arbitrary array key which can be omitted as it has no impact on functionality, I updated the readme for clarity (note that unit tests do not use the named keys)
`Title` is the name of the field which this feature doesn't dictate

I'm don't think this feature is complex, but it probably needs improved documentation. It's really just a simple blacklist / whitelist for fields removal. You can target properties and fields.

Namespacing convention used is feature-based namespace. This approach is much more maintainable long term compared to the type-based namespace (mostly inherited from SS 3).

Feature namespacing works like this:

- namespace is named after feature, not type (for example `MyFeature`)
- feature is a folder which:
- has files named based on type (for example `Model`, `Extension`, `Service`...)
- has folders which represent sub-features

When you use a class from feature namespace you simply import the namespace of the feature rather than the specific file.

For example:

`use SomeOtherStuff\MyFeature;`

And then it you code you would use specific classes like this:

`MyFeature\Model`
`MyFeature\Extension`
`MyFeature\Service`

The advantage is that the files are structured in a way where you naturally would want them because most of the time developers are working on features rather than types.

The type based namespace on the other hand often requires the developer to open several different type folders to get the files that are desired.

For the general approach, I agree that changing the field scaffolder may be a good option from feature point of view. It's actually what I started working on originally, I changed the hard coded scaffolder configuration to configuration API, however this approach has a big problem.

There are so many instances where module code assumes that certain fields will be scaffolded, it's way too much effort to update all that code. If you remove these fields, the app will break as the code doesn't bother to check for fields availability.

On the other hand, removing scaffolded fields later via an extension has the added flexibility to control when fields are removed. Often times other modules add additional fields which are undesired and this wouldn't be covered by changing the scaffolder.

kinglozzer commented:
I’m not convinced this solution is better than our existing API of defining fields in PHP to be honest. Developers now have another place to look (model PHP files, extension PHP files and now YAML config files) to work out why a given field as added/omitted, and given the additional YAML required I don’t see it saving developers much time/effort. It might save lines of code, but in terms of how easy it is to understand what’s happening at a glance I think this would be a step back.

I can see you’ve spent a reasonable amount of time writing this and writing tests to cover it, so I feel bad for saying that IMO we shouldn’t merge this 😅. Maybe in future it’d be best to discuss potential solutions in the issue before spending your time on implementing it (I know you work for SS Ltd, so maybe you’ve already discussed this internally!)

Regarding class names/namespacing, I can see some of the merits of feature based namespacing but it does go against the current conventions in core code so I agree that a more descriptive base class name would be best here

mfendeksilverstripe commented:
Thanks for feedback. I agree that this solution has some issues, however the fact is that on every project I worked on I implemented this extension over and over again. That's the opposite of DRY. 

I don't think this is wasted effort because even if this PR doesn't get merged it raises the awareness of this issue and I can always move it to my team's module and reuse the code anyway.

I disagree on the point with namespaces. Consistency with existing code is not more important than better code structure. We use feature-based namespaces in our team on new projects, but also older existing ones which have the majority of code base in the old type based namespace.

Having a mixed namespace approach in your code is not an issue as the class names are almost completely identical. For example a class which was mentioned above would look like this

`DisableScaffoldingExtension` in the type base namespace

and like this

`DisableScaffolding\Extension` in the feature base namespace

The difference is small enough that the inconsistency doesn't disrupt the development very much.

Of course, this is just a suggestion as it's up to the module owner to decide the namespace type.

michalkleiner commented:
> Having a mixed namespace approach in your code is not an issue as the class names are almost completely identical. For example a class which was mentioned above would look like this
> 
> `DisableScaffoldingExtension` in the type base namespace
> 
> and like this
> 
> `DisableScaffolding\Extension` in the feature base namespace

Using the class in code via the feature base namespace requires the developer to use that approach deliberately, the type base namespace doesn't, therefore it's IMHO less prone to "not following a convention".

I'd suggest creating an RFC issue to establish a convention going forward to avoid two or more different approaches in the core code?

mfendeksilverstripe commented:
@michalkleiner Sounds good, thanks :) Will follow up on that.

robbieaverill commented:
Similar to the discussion previously about protected versus private methods, this is another example of something that we kind of have a de facto standard for by virtue of age, but I agree it would be good for contributors if we clarified it in docs

mfendeksilverstripe commented:
Moved to a separate module https://github.com/silverstripe-terraformers/scaffolded-fields
